### DISCUSSION TITLE
INAV, airplane, Alt Hold, Pos Hold, RTH problem, goes into spin.

### DISCUSSION THREAD

**Discussion by KaptenAndreas** [iNavFlight/inav#6392] [Q&A]
## Current Behavior

I have problems with Altitude Hold, Pos Hold, RTH.
It is a new plane Ranger G2 with INAV on a Matek 765, GPS from Matek M8Q. I have done a Servo trim and Autotune this went well. It flies nice in AIR, Angle, Horizon and also Manual mode. Accelerometer and compass have been calibrated. The 765 FC is mounted rotated 180 degree in yaw axis setup in INAV made.

Now I want to try some more flight modes.
When flying in Angle or Horizon mode that feels ok flies as expected, nice and smooth. I now activate Altitude hold. It holds the altitude great, now I try to turn with roll left or right and after say 10 degree of bank angle then the plane goes into a roll and drops altitude quickly and is out of control. I deactivate AH and I get control back. I try again, fly straight in AH and now try to turn left or right with yaw same problem at about 10 degree bank it rolls out of control. See zipped file with dvr recording.

I tried to turn the plane really really slowly not practical but for test then it holds altitude and dont go out of control. But as soon I am a little bit more to a normal turn bank it is out of control.

I try to activate Pos Hold or RTH and get the same problem.

What can be the problem? Does somebody have some good suggestions?

I have a short dvr film and also my dump and log file from the FC.

## Steps to Reproduce

1. See decription above.
2.
3.
4.

## Expected behavior
 
It should be possible to bank and change flight direction and at the same time keep altitude. Same thing when using Pos Hold or RTH

## Additional context

[201205RangerG2Dump.txt](https://github.com/iNavFlight/inav/files/5654825/201205RangerG2Dump.txt)

[LOG00018.TXT](https://github.com/iNavFlight/inav/files/5654837/LOG00018.TXT)

[InavRangerG2v2.zip](https://github.com/iNavFlight/inav/files/5654887/InavRangerG2v2.zip)

---

- FC Board name and vendor: Matek 765
- INAV version string: # version
# INAV/MATEKF765 2.5.2 Aug  4 2020 / 10:30:32 (faaedc7cb)
# GCC-9.2.1 20191025 (release) [ARM/arm-9-branch revision 277599]

[a comment by an automated account is not shown]

**Comment by avsaase** [comment]
I have had this exact same problem, with the same airplane so I think I can help. I will assume you have correctly setup the rates, also for yaw (I know, hard to measure correctly) and that auto tune has given you reasonable P, I and FF gains. Also check that your rudder throws in manual mode are sensible, if they are too large it is best to reduce them mechanically. Alternatively you can reduce the yaw FF gain. If these things are not setup correctly you should fix those things first. There are some great guides on https://inavfixedwinggroup.com/.

Having said that, the problem is that the rudder on this airplane is incredibly effective, and INAV is overusing the rudder when making turns in the navigation modes. On this airplane rudder and ailerons are very much cross-controlled, but INAV doesn't understand this. So when INAV tries to use rudder to create a yaw rotation to make a nice coordinated turn, it is actually adding a lot of roll. This causes the plane to tip into a spin and also explains why switching to any non-navigation mode immediately recovers from the spin; these modes do not automatically apply rudder. You can reduce the amount of rudder used for coordinated turns by decreasing `fw_turn_assist_yaw_gain` from the default of 1. If you set it to zero it will not use any rudder, but then the rudder might actually work the other way to fight the turn (something I am trying to solve in #6387). 

TL;DR: if everything else is **really** setup correctly, reduce `fw_turn_assist_yaw_gain`.

**Comment by KaptenAndreas** [comment]
> I have had this exact same problem, with the same airplane so I think I can help. I will assume you have correctly setup the rates, also for yaw (I know, hard to measure correctly) and that auto tune has given you reasonable P, I and FF gains. Also check that your rudder throws in manual mode are sensible, if they are too large it is best to reduce them mechanically. Alternatively you can reduce the yaw FF gain. If these things are not setup correctly you should fix those things first. There are some great guides on https://inavfixedwinggroup.com/.
> 
> Having said that, the problem is that the rudder on this airplane is incredibly effective, and INAV is overusing the rudder when making turns in the navigation modes. On this airplane rudder and ailerons are very much cross-controlled, but INAV doesn't understand this. So when INAV tries to use rudder to create a yaw rotation to make a nice coordinated turn, it is actually adding a lot of roll. This causes the plane to tip into a spin and also explains why switching to any non-navigation mode immediately recovers from the spin; these modes do not automatically apply rudder. You can reduce the amount of rudder used for coordinated turns by decreasing `fw_turn_assist_yaw_gain` from the default of 1. If you set it to zero it will not use any rudder, but then the rudder might actually work the other way to fight the turn (something I am trying to solve in #6387).
> 
> TL;DR: if everything else is **really** setup correctly, reduce `fw_turn_assist_yaw_gain`.

Thanks for the tips I will double check my rudder travel motion and if necessary maybe reduce them a little. And then test to change `fw_turn_assist_yaw_gain`.

**Comment by KaptenAndreas** [comment]
> I have had this exact same problem, with the same airplane so I think I can help. I will assume you have correctly setup the rates, also for yaw (I know, hard to measure correctly) and that auto tune has given you reasonable P, I and FF gains. Also check that your rudder throws in manual mode are sensible, if they are too large it is best to reduce them mechanically. Alternatively you can reduce the yaw FF gain. If these things are not setup correctly you should fix those things first. There are some great guides on https://inavfixedwinggroup.com/.
> 
> Having said that, the problem is that the rudder on this airplane is incredibly effective, and INAV is overusing the rudder when making turns in the navigation modes. On this airplane rudder and ailerons are very much cross-controlled, but INAV doesn't understand this. So when INAV tries to use rudder to create a yaw rotation to make a nice coordinated turn, it is actually adding a lot of roll. This causes the plane to tip into a spin and also explains why switching to any non-navigation mode immediately recovers from the spin; these modes do not automatically apply rudder. You can reduce the amount of rudder used for coordinated turns by decreasing `fw_turn_assist_yaw_gain` from the default of 1. If you set it to zero it will not use any rudder, but then the rudder might actually work the other way to fight the turn (something I am trying to solve in #6387).
> 
> TL;DR: if everything else is **really** setup correctly, reduce `fw_turn_assist_yaw_gain`.

Hi again!

One more question to your answer.
Should I set the correct rates for the rudders in Manual Mode or in Angle/Horizon mode?
Because there is a big difference between them.

/KaptenAndreas

**Comment by avsaase** [comment]
The rates I referred to, `roll_rate`, `pitch_rate` and `yaw_rate`, should be set to the rate of rotation at full servo deflection. In other words, the reflect the maximum rotation rate the airplane is capable of. You need to measure this in manual, either by frame-by-frame playing a DVR or by recording a blackbox log. I prefer the last method because it is easier to measure the yaw rate that way.
